File: kdb-document/src/commonMain/kotlin/dev/kdb/document/KdbOp.kt
package dev.kdb.document

import dev.kdb.codec.KdbHash
import dev.kdb.codec.KdbUuid
import dev.kdb.codec.KdbValue
import dev.kdb.codec.encodeToBytes
import dev.kdb.codec.toUuidVal
import dev.kdb.codec.toKdbUuid
import dev.kdb.document.internal.sha256Digest

/**
 * Atomic unit of change within a transaction.
 */
public sealed class KdbOp {
    public data class Write(
        val docId: KdbUuid,
        val patch: String,
    ) : KdbOp()

    public data class Delete(
        val docId: KdbUuid,
    ) : KdbOp()

    public data class FileWrite(
        val path: String,
        val blobHash: KdbHash,
    ) : KdbOp()

    public data class SchemaMigration(
        val migrationId: KdbUuid,
        val migrationPayload: String,
    ) : KdbOp()

    public fun toKdbValue(): KdbValue =
        when (this) {
            is Write ->
                KdbValue.UnionVal(
                    0,
                    KdbValue.RecordVal(
                        mapOf(
                            1 to docId.toUuidVal(),
                            2 to KdbValue.StringVal(patch),
                        ),
                    ),
                )
            is Delete ->
                KdbValue.UnionVal(
                    1,
                    KdbValue.RecordVal(
                        mapOf(1 to docId.toUuidVal()),
                    ),
                )
            is FileWrite ->
                KdbValue.UnionVal(
                    2,
                    KdbValue.RecordVal(
                        mapOf(
                            1 to KdbValue.StringVal(path),
                            2 to KdbValue.FixedVal(blobHash.bytes.copyOf()),
                        ),
                    ),
                )
            is SchemaMigration ->
                KdbValue.UnionVal(
                    3,
                    KdbValue.RecordVal(
                        mapOf(
                            1 to migrationId.toUuidVal(),
                            2 to KdbValue.StringVal(migrationPayload),
                        ),
                    ),
                )
        }

    public companion object {
        public fun fromKdbValue(value: KdbValue): KdbOp {
            val uv = value as? KdbValue.UnionVal ?: throw CommitDecodeException("KdbOp: expected union")
            return when (uv.branch) {
                0 -> {
                    val r = uv.value as? KdbValue.RecordVal ?: throw CommitDecodeException("OpWrite record")
                    val id =
                        (r.fields[1] as? KdbValue.UuidVal)?.toKdbUuid()
                            ?: throw CommitDecodeException("OpWrite docId")
                    val patch =
                        (r.fields[2] as? KdbValue.StringVal)?.v
                            ?: throw CommitDecodeException("OpWrite patch")
                    Write(id, patch)
                }
                1 -> {
                    val r = uv.value as? KdbValue.RecordVal ?: throw CommitDecodeException("OpDelete record")
                    val id =
                        (r.fields[1] as? KdbValue.UuidVal)?.toKdbUuid()
                            ?: throw CommitDecodeException("OpDelete docId")
                    Delete(id)
                }
                2 -> {
                    val r = uv.value as? KdbValue.RecordVal ?: throw CommitDecodeException("OpFileWrite record")
                    val path =
                        (r.fields[1] as? KdbValue.StringVal)?.v
                            ?: throw CommitDecodeException("OpFileWrite path")
                    val fh =
                        (r.fields[2] as? KdbValue.FixedVal)?.v
                            ?: throw CommitDecodeException("OpFileWrite blobHash")
                    FileWrite(path, KdbHash.fromBytes(fh.copyOf()))
                }
                3 -> {
                    val r = uv.value as? KdbValue.RecordVal ?: throw CommitDecodeException("OpSchemaMigration record")
                    val mid =
                        (r.fields[1] as? KdbValue.UuidVal)?.toKdbUuid()
                            ?: throw CommitDecodeException("OpSchemaMigration migrationId")
                    val payload =
                        (r.fields[2] as? KdbValue.StringVal)?.v
                            ?: throw CommitDecodeException("OpSchemaMigration payload")
                    SchemaMigration(mid, payload)
                }
                else -> throw CommitDecodeException("unknown KdbOp branch ${uv.branch}")
            }
        }
    }
}

File: go/kdb/document/kdb_op.go
package document

import (
	"github.com/limidus/kdb/go/kdb/codec"
)

// Op is an atomic change within a transaction.
type Op interface {
	isOp()
	toValue() codec.Value
}

type WriteOp struct {
	DocID codec.UUID
	Patch string
}

func (WriteOp) isOp() {}

func (o WriteOp) toValue() codec.Value {
	return codec.UnionValue{
		Branch: 0,
		Inner: codec.RecordValue{Fields: map[int]codec.Value{
			1: uuidVal(o.DocID),
			2: codec.StringValue{V: o.Patch},
		}},
	}
}

type DeleteOp struct {
	DocID codec.UUID
}

func (DeleteOp) isOp() {}

func (o DeleteOp) toValue() codec.Value {
	return codec.UnionValue{
		Branch: 1,
		Inner: codec.RecordValue{Fields: map[int]codec.Value{
			1: uuidVal(o.DocID),
		}},
	}
}

type FileWriteOp struct {
	Path     string
	BlobHash codec.Hash
}

func (FileWriteOp) isOp() {}

func (o FileWriteOp) toValue() codec.Value {
	return codec.UnionValue{
		Branch: 2,
		Inner: codec.RecordValue{Fields: map[int]codec.Value{
			1: codec.StringValue{V: o.Path},
			2: hashVal(o.BlobHash),
		}},
	}
}

type SchemaMigrationOp struct {
	MigrationID      codec.UUID
	MigrationPayload string
}

func (SchemaMigrationOp) isOp() {}

func (o SchemaMigrationOp) toValue() codec.Value {
	return codec.UnionValue{
		Branch: 3,
		Inner: codec.RecordValue{Fields: map[int]codec.Value{
			1: uuidVal(o.MigrationID),
			2: codec.StringValue{V: o.MigrationPayload},
		}},
	}
}

// OpFromValue decodes a wire KdbOp union value.
func OpFromValue(value codec.Value) (Op, error) {
	uv, ok := value.(codec.UnionValue)
	if !ok {
		return nil, NewCommitDecodeError("KdbOp: expected union", nil)
	}
	rec, ok := uv.Inner.(codec.RecordValue)
	if !ok {
		return nil, NewCommitDecodeError("KdbOp: expected record", nil)
	}
	switch uv.Branch {
	case 0:
		id, err := uuidFromVal(rec.Fields[1])
		if err != nil {
			return nil, NewCommitDecodeError("OpWrite docId", nil)
		}
		patch, ok := rec.Fields[2].(codec.StringValue)
		if !ok {
			return nil, NewCommitDecodeError("OpWrite patch", nil)
		}
		return WriteOp{DocID: id, Patch: patch.V}, nil
	case 1:
		id, err := uuidFromVal(rec.Fields[1])
		if err != nil {
			return nil, NewCommitDecodeError("OpDelete docId", nil)
		}
		return DeleteOp{DocID: id}, nil
	case 2:
		path, ok := rec.Fields[1].(codec.StringValue)
		if !ok {
			return nil, NewCommitDecodeError("OpFileWrite path", nil)
		}
		h, err := hashFromVal(rec.Fields[2])
		if err != nil {
			return nil, err
		}
		return FileWriteOp{Path: path.V, BlobHash: h}, nil
	case 3:
		id, err := uuidFromVal(rec.Fields[1])
		if err != nil {
			return nil, NewCommitDecodeError("OpSchemaMigration migrationId", nil)
		}
		payload, ok := rec.Fields[2].(codec.StringValue)
		if !ok {
			return nil, NewCommitDecodeError("OpSchemaMigration payload", nil)
		}
		return SchemaMigrationOp{MigrationID: id, MigrationPayload: payload.V}, nil
	default:
		return nil, NewCommitDecodeError("unknown KdbOp branch", nil)
	}
}
